File: old_code/src_python/nmpccodegen/Cfunctions/IndicatorBoxFunction.py
from .ProximalFunction import ProximalFunction
from .Function import Cfunction
from .source_file_operations import Source_file_generator


class IndicatorBoxFunctionProx(Cfunction):
    def __init__(self,lower_limits,upper_limits):
        """
        constructs box function

        Parameters
        ---------
        lower_limit : array of the lowest values the input can be
        upper_limit : array of the highest value the input can be
        """
        self._lower_limits=lower_limits
        self._upper_limits=upper_limits
        self._dimension=min(len(upper_limits),len(lower_limits))
        if(len(lower_limits)!=len(upper_limits)):
            print("Error limits length do not match, using the smallest of the two: dimension=" \
                  +str(self._dimension))

    def generate_c_code(self, location):
        """
        Generates C code of proximal function

        Parameters
        ---------
        location : location function needs to be generated
        """
        source_file = Source_file_generator(location,"proxg")
        source_file.open()
        source_file.start_for("i","MPC_HORIZON",indent=1)

        for dimension in range(0,self._dimension):
            source_file.write_comment_line(\
                "check if the value of the border is outside the box, if so go to the nearest point inside the box", \
                indent=2)
            source_file.write_line("if(state["+str(dimension)+"]<"+str(self._lower_limits[dimension])+"){",indent=2)
            source_file.set_output(dimension,str(self._lower_limits[dimension]),3)
            source_file.write_line("}else if(state[" + str(dimension) + "]>" + str(self._upper_limits[dimension]) + "){",
                                   indent=2)
            source_file.set_output(dimension, str(self._upper_limits[dimension]), 3)
            source_file.write_line("}else{", 2)
            source_file.set_output(dimension, "state[" + str(dimension) + "]", 3)
            source_file.write_line("}", 2)

        source_file.write_line("state+="+str(self._dimension)+ ";",2)
        source_file.close_for(indent=1)
        source_file.close()

class IndicatorBoxFunction(ProximalFunction):
    def __init__(self,lower_limits,upper_limits):
        super().__init__(IndicatorBoxFunctionProx(lower_limits, upper_limits))
        """constructor box function"""
        self._lower_limits=lower_limits
        self._upper_limits=upper_limits
        self._dimension=min(len(upper_limits),len(lower_limits))
        if(len(lower_limits)!=len(upper_limits)):
            print("Error limits length do not match, using the smallest of the two: dimension=" \
                  +str(self._dimension))

    def generate_c_code(self, location):
        source_file = Source_file_generator(location,"g")
        source_file.open()
        source_file.start_for("i","MPC_HORIZON",indent=1)

        for dimension in range(0,self._dimension):
            source_file.write_comment_line("check if the value of the border is outside the box, if so return zero", \
                                           indent=1)
            source_file.write_line("if(state["+str(dimension)+"]<"+str(self._lower_limits[dimension])+ \
                              " || state["+str(dimension)+"]>"+str(self._upper_limits[dimension])+ \
                              "){",indent=1)
            source_file.write_line("return LARGE;", indent=2)
            source_file.write_line("}",indent=1)

        source_file.write_line("state+="+str(self._dimension)+ ";",2)
        source_file.close_for(indent=1)

        source_file.write_comment_line("if the value's where never outside the box, return zero",indent=1)
        source_file.write_line("return 0;", indent=1)
        source_file.close()

def main():
    test = IndicatorBoxFunction([1,2],[3,4])
    test.generate_c_code("test.c")

if __name__ == "__main__":
    main()

File: old_code/src_matlab/+nmpccodegen/+Cfunctions/@IndicatorBoxFunction/IndicatorBoxFunction.m
classdef IndicatorBoxFunction < nmpccodegen.Cfunctions.ProximalFunction
    %INDICATORBOXFUNCTION An indicator Box function that can be used as
    %input constraint.
    %   - 
    
    properties
        lower_limits %  the lower limits of the box.
        upper_limits % The upper limits of the box.
        dimension % Dimension of the box, this is equal to the length of the array's lower_limits and upper_limits.
    end
    
    methods
        function obj=IndicatorBoxFunction(lower_limits,upper_limits)
            % lower_limits :  the lower limits of the box.
            % upper_limits : The upper limits of the box.
            
            obj.lower_limits=lower_limits;
            obj.upper_limits=upper_limits;
            obj.dimension = min(length(lower_limits),length(upper_limits));
            
            obj.prox =  nmpccodegen.Cfunctions.IndicatorBoxFunctionProx(lower_limits,upper_limits);
        end
        function obj=generate_c_code(obj, location)
            % Generate the C code at the folder sepcified in location.
            
            source_file = nmpccodegen.Cfunctions.Source_file_generator(location,'g');
            source_file = source_file.open();
            source_file = source_file.start_for('i','MPC_HORIZON',1);
            
            for i_dimension=1:obj.dimension
                source_file.write_comment_line('check if the value of the border is outside the box, if so return zero', ...
                                               2);
                source_file = source_file.write_line( ['if(state[' num2str(i_dimension-1) ...
                    ']<' num2str(obj.lower_limits(i_dimension)) ...
                    ' || state[' num2str(i_dimension-1) ']>' ... 
                    num2str(obj.upper_limits(i_dimension))  ...
                    '){'],2);
                source_file = source_file.write_line('return LARGE;',3);
                source_file = source_file.write_line('}',2);
            end
            
            source_file = source_file.write_line(['state+=' num2str(obj.dimension) ';'],2);
            source_file = source_file.close_for(1);

            source_file = source_file.write_comment_line('if the values where never outside the box, return zero',1);
            source_file = source_file.write_line( 'return 0;', 1);
            source_file.close();
        end
    end
    
end